pkg/proxy/iptables/proxier.go

```diff
@@ -1448,21 +1479,17 @@ func (proxier *Proxier) syncProxyRules() {
 			"-m", "comment", "--comment", fmt.Sprintf(`"route LOCAL traffic for %s LB IP to service chain"`, svcNameString),
 			"-m", "addrtype", "--src-type", "LOCAL", "-j", string(svcChain))...)
 
-		numLocalEndpoints := len(localEndpointChains)
-		if numLocalEndpoints == 0 {
-			// Blackhole all traffic since there are no local endpoints
-			args = append(args[:0],
-				"-A", string(svcXlbChain),
-				"-m", "comment", "--comment",
-				fmt.Sprintf(`"%s has no local endpoints"`, svcNameString),
-				"-j",
-				string(KubeMarkDropChain),
-			)
-			writeLine(proxier.natRules, args...)
-		} else {
+		// This is where we jump to local endpoint chains. The list of local endpoints is selected based on the first
+		// set of endpoints that are one of:
+		//   1. ready and NOT terminating
+		//   2. ready and terminating
+		//   3. otherwise, blackhole traffic
+		numLocalReadyEndpoints := len(localReadyEndpointChains)
+		numLocalServingTerminatingEndpoints := len(localServingTerminatingEndpointChains)
+		if numLocalReadyEndpoints > 0 {
```

Comment:
@robscott With Slice subsetting, do we guarantee that a given node always has endpoints on that same node in its subset, when either internal or external traffic policy is in play?  We'll need to make sure we do.

Reply:
subsetting is per zone, so as long as kube-proxy is consuming endpointslices for the appropriate zone this should not be an issue. Of course if we wanted to further subset slices we'd need to take each node into account.

Reply:
I thought it was possible, in case of imbalance, for zone A to "donate" endpoints to the for-zone=B slice(s).  In that case the node can't access it's own pod and this will break?


E.g.

Pod p lands on node n in zone z1
Service s is OnlyLocal
Service s selects Pod p (among others)
The for-zone=z1 slice has too many endpoint
The for-zone=z2 slice has too few
Donate p from for-zone=z1 to z2
Traffic thru n1 (which hosts p) does not have p as an endpoint
Pod p never gets any traffic

Is that real?  Or am I missing some detail that prevents it?

Reply:
You're completely right. For OnlyLocal we should not use any kind of advanced subsetting. We can either disable the feature or use an approach like we'd proposed with `Require` where each endpoint would get assigned to a slice for the zone it was in.

Reply:
Good opportunity fior test cases!! :)

Reply:
https://github.com/kubernetes/enhancements/pull/2153